### PR TITLE
Blog post on un-mixing a mixin

### PR DESCRIPTION
**Pull request by mirisuzanne**
## Description

Something I was thinking about after CSS Day
https://deploy-preview-614--oddleventy.netlify.app/2024/06/11/removing-mixins/

## Related Issue(s)

Should I make more mention of CSS Day, or my role there, or the workshop (somehow?)

### CODE REVIEW

File: content/blog/2024/removing-mixins.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+[CSS-native mixins](https://css.oddbird.net/sasslike/mixins-functions/).
+How do we create re-usable blocks of styling
+that can be 'mixed in' to various selectors,
+based on arbitrary conditions?
+
+## Mixin substitution with `@apply`
+
+I made [a proposal](https://css.oddbird.net/sasslike/mixins-functions/)
+last year, and it was
+[adopted by the CSS Working Group](https://github.com/w3c/csswg-drafts/issues/9350#issuecomment-1939628591),
```
Comment (dvdherron) [inline content/blog/2024/removing-mixins.md:24]:
```suggestion
[adopted by the CSS Working Group](https://github.com/w3c/csswg-drafts/issues/9350#issuecomment-1939628591)
```

File: content/blog/2024/removing-mixins.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+  }
+}
+
+/* apply the mixin by giving --fancy-em a value */
+p {
+  --fancy-em: mediumvioletred, teal;
+}
+```
+
+(you can see this [style query demo](https://codepen.io/miriamsuzanne/pen/qBGXMPg?editors=1100) working in a Chromium browser)
```
Comment (dvdherron) [inline content/blog/2024/removing-mixins.md:121]:
```suggestion
(You can see this [style query demo](https://codepen.io/miriamsuzanne/pen/qBGXMPg?editors=1100) working in a Chromium browser.)
```

File: content/blog/2024/removing-mixins.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+  We intentionally avoided cascading mixin _definitions_ --
+  one mixin name can't refer to different things in different places.
+  Would cascading `@apply` rules have similar issues?
+
+I don't think we want something like
+the JavaScript `removeEventListener()` function,
+which requires a careful matching of arguments.
+I'd like to avoid any `@un-apply`-style rules.
+That has always seemed fragile to me,
+and I'd rather use the cascade mechanic if we can.
```
Comment (dvdherron) [inline content/blog/2024/removing-mixins.md:371]: should "mechanic" be "mechanism"?

**Review by dvdherron — APPROVED** [review]

File: content/blog/2024/removing-mixins.md (hunk 3 above)
Comment (mirisuzanne) [reply content/blog/2024/removing-mixins.md:371]: Maybe? Or it might work without that word entirely?

Comment (mirisuzanne) [reply content/blog/2024/removing-mixins.md:371]:
```suggestion
and I'd rather use the cascade if we can.
```

File: content/blog/2024/removing-mixins.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+are they intertwined?
+If so, they belong in the same property.
+
+It seems clear to me that it would be
+useful (at least sometimes)
+for the mixin-application syntax to cascade.
+But before we make any big changes
+we also need to ask:
+
+- Are there place we _don't want_ that behavior?
```
Comment (jgerigmeyer) [inline content/blog/2024/removing-mixins.md:356]:
```suggestion
- Are there places we _don't want_ that behavior?
```

File: content/blog/2024/removing-mixins.md [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+I don't think we want something like
+the JavaScript `removeEventListener()` function,
+which requires a careful matching of arguments.
+I'd like to avoid any `@un-apply`-style rules.
+That has always seemed fragile to me,
+and I'd rather use the cascade if we can.
+
+What do you think?
+Should mixin-calls cascade?
```
Comment (jgerigmeyer) [inline content/blog/2024/removing-mixins.md:374]:
```suggestion
Should mixin calls cascade?
```

File: content/blog/2024/removing-mixins.md [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+Are there use-cases for both behaviors?
```
Comment (jgerigmeyer) [inline content/blog/2024/removing-mixins.md:375]:
```suggestion
Are there use cases for both behaviors?
```

**Review by jgerigmeyer — APPROVED** [review]
Needs an image, but otherwise LGTM 👍

**Review by SondraE — APPROVED** [review]
Should _Roma_ Komarov should be _Roman_ throughout the article? That's what's on their website when you click through, anyhow.

I don't think "off-switch" needs a hyphen?

You end the article with questions. You could add something like: _come discuss this with me at the upcoming Cascading Layout Workshop on June 24. You can register below._ 
Additionally you could put the registration form CTA near the beginning of the article somewhere with an invitation something like: _If you questions about all this, come talk about it with me live on June 24-26._

**Comment by mirisuzanne** [comment]
> Should _Roma_ Komarov should be _Roman_ throughout the article? That's what's on their website when you click through, anyhow.

We could. He uses Roma on Mastodon - just the short form of the name. I think either is fine if we're consistent?

> You end the article with questions. You could add something like: _come discuss this with me at the upcoming Cascading Layout Workshop on June 24. You can register below._
> Additionally you could put the registration form CTA near the beginning of the article with an invitation something like: _If you have questions about all this, come ask them. I'll be doing live Q&A at the Cascade Layouts Workshop June 24-26._

Does it matter that it's a layout workshop, and this post isn't really layout related at all?

File: content/blog/2024/removing-mixins.md [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+  @apply --visually-hidden;
+}
+```
+
+Or we need to carefully revert
+every property of the mixin.
+But… revert _to what value_?
+Maybe we can use `revert-layer`
+and some clever layering?
+We could even build the off-switch into our mixin:
```
Comment (mirisuzanne) [inline content/blog/2024/removing-mixins.md:300]:
```suggestion
We could even build the off switch into our mixin:
```

File: content/blog/2024/removing-mixins.md [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+@layer overrides {
+  :focus,
+  :active {
+    /* set --off to revert-layer */
+    @apply --visually-hidden(revert-layer);
+  }
+}
+```
+
+That provides an off-switch,
```
Comment (mirisuzanne) [inline content/blog/2024/removing-mixins.md:330]:
```suggestion
That provides an off switch,
```

File: content/blog/2024/removing-mixins.md [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+{{ embed.codepen(
+  id='dyEzQPz',
+  title='Un-mixing a mixin',
+  user='miriamsuzanne',
+  tab='css,result'
+) }}
+
+
+## It's hard to un-mix a previously mixed-in mixin once mixed
```
Comment (stacyk) [inline content/blog/2024/removing-mixins.md:217]: Great heading

**Review by stacyk — APPROVED** [review]
Great article. Great examples in it as well. Approving but I don't think it has a hero image yet?

**Comment by mirisuzanne** [comment]
@SondraE posted to masto:

- https://front-end.social/@mia/112599109654303752

**Comment by SondraE** [comment]
@mirisuzanne @stacyk @dvdherron 
OddBird boosted on masto and posted on:
https://www.linkedin.com/feed/update/urn:li:activity:7206377634627264512/
https://x.com/OddBird/status/1800614164064358555

**Pull request merged**